src/modules/bootloader/main.py

```diff
@@ -145,43 +127,84 @@ def create_systemd_boot_conf(install_path, efi_dir, uuid, entry, entry_name, ker
         kernel_params.append("resume=/dev/mapper/{!s}".format(
             swap_outer_mappername))
 
-    kernel_line = get_kernel_line(kernel_type)
+    kernel_line = entry_name + " " + kernel_version
     libcalamares.utils.debug("Configure: \"{!s}\"".format(kernel_line))
 
     if kernel_type == "fallback":
-        img = libcalamares.job.configuration["fallback"]
-        entry_name = entry_name + "-fallback"
+        version_string = kernel_version + "-fallback"
+        initrd = "initrd-fallback"
+        mkinitcpio_option = "-S autodetect"
     else:
-        img = libcalamares.job.configuration["img"]
+        version_string = kernel_version
+        initrd = "initrd"
+        mkinitcpio_option = ""
 
-    conf_path = os.path.join(install_path + efi_dir,
-                             "loader",
-                             "entries",
-                             entry_name + ".conf")
+    # get the machine-id
+    with open(os.path.join(install_path, "etc", "machine-id"), 'r') as machineid_file:
+        machine_id = machineid_file.read().rstrip('\n')
 
-    # Copy kernel and initramfs to a subdirectory of /efi partition
-    files_dir = os.path.join(install_path + efi_dir, entry_name)
-    os.mkdir(files_dir)
+    # Copy kernel to a subdirectory of /efi partition
+    machine_dir = os.path.join(install_path + efi_dir, machine_id)
+    try:
+        os.mkdir(machine_dir)
+    except FileExistsError: # We can ignore errors caused by the directory existing already
+        pass
 
-    kernel_path = install_path + kernel
+    files_dir = os.path.join(machine_dir, kernel_version)
+    try:
+        os.mkdir(files_dir)
+    except FileExistsError: # We can ignore errors caused by the directory existing already
+        pass
+    
+    kernel_path = os.path.join(install_path, kernel)
     kernel_name = os.path.basename(kernel_path)
-    shutil.copyfile(kernel_path, os.path.join(files_dir, kernel_name))
-
-    img_path = install_path + img
-    img_name = os.path.basename(img_path)
-    shutil.copyfile(img_path, os.path.join(files_dir, img_name))
+    shutil.copyfile(kernel_path, os.path.join(files_dir, "linux"))
 
+    # generate the initramfs - this is Arch specific and should be replaced for other distros
```

Comment:
I suppose it isn't Arch specific.  It is specific to a distro using both `mkinitcpio` and `kernel-install` which is still fairly specific.

It doesn't have to be done here.  It could be done before or after this point.  It needs the kernels to be installed and the machine id to be generated.

As for using the existing initcpio module, the problem would be that it would need to be refactored substantially in order to add the ability to search for all the kernels to know where to write the initrd.  In c++, that would be a decent amount of code.  It would also mean all that logic was duplicated in two places, once in c++ and once in python.

I am not broadly familiar with Calamares outside of this module.  Is there a better way to approach it?